src/qibolab/_core/instruments/emulator/emulator.py

```diff
@@ -94,29 +96,47 @@ def _single_sequence(
         assert options.nshots is not None
         results = {}
         levels = np.array(configs["hamiltonian"].transmon_levels)
-        for ro_id in self._acquisitions(sequence):
-            # TODO: find better to extract qubit from acquisition channel
-            # possible solution _single_sequence should be aware of platform.qubits
-            qubit = int(sequence.pulse_channels(ro_id)[0][0])
-            single_qubit_probabilities = np.moveaxis(
-                np.moveaxis(probabilities, -1, 0)[
-                    qubit * levels : (qubit + 1) * levels
-                ],
-                0,
-                -1,
+        # store density matrices for each duration for the whole system
+        density_matrices = {}
+
+        probabilities = np.moveaxis(probabilities, -2, 0)
+        for i, (ro_id, sample) in enumerate(self._acquisitions(sequence).items()):
+            density_matrices.setdefault(
+                sample,
+                shots(probabilities[i], options.nshots)
+                if options.averaging_mode == AveragingMode.SINGLESHOT
+                else probabilities[i],
             )
+        for ro_id, sample in self._acquisitions(sequence).items():
+
+            def _get_qubit(a: int, qubit: int, n: int):
+                return a // n if qubit == 0 else a % n
 
+            def _get_probability(probs: list, qubit: int, n: int):
+                states = np.arange(n)
+                probs = probs.reshape(n, n)
+                if qubit == 1:
+                    return sum(np.dot(states, probs.T))
+                elif qubit == 0:
+                    return sum(np.dot(states, probs))
```

Comment:
Avoid duplication as much as possible
```suggestion
                return sum(np.dot(states, probs.T if qubit == 1 else probs))
```

Moreover, the `if ... elif` pattern leaves room for returning `None` in case of unexpected input. If you intend this function to be always used with `0` and `1` inputs, `assert` them.
In general, you should try to avoid validating user input as much as possible, possibly relying on static validation. However, here `qubit` is declared to be an `int`, while in practice only `0` and `1` are acceptable values. Better to `assert`.